libraries/chain/include/eosio/chain/chain_config.hpp

```diff
@@ -184,6 +184,60 @@ struct chain_config_v1 : chain_config_v0 {
    }
 };
 
+/**
+ * @brief v2 Producer-voted blockchain configuration parameters
+ *
+ * If Adding new parameters create chain_config_v[n] class instead of adding
+ * new parameters to v2, v1 or v0. This is needed for snapshots backward compatibility
+ */
+struct chain_config_v2 : chain_config_v1 {
+   using Base = chain_config_v1;
+
+   uint32_t   max_sync_call_depth = config::default_max_sync_call_depth;            ///< size limit for sync call depth
+   uint32_t   max_sync_call_data_size = config::default_max_sync_call_data_size;    ///< size limit for sync call data: input and return value respectively
+
+   //order must match parameters as ids are used in serialization
+   enum {
+     max_sync_call_depth_id = Base::PARAMS_COUNT,
+     max_sync_call_data_size_id,
+     PARAMS_COUNT
+   };
+
+   inline const Base& base() const {
+      return static_cast<const Base&>(*this);
+   }
+
+   void validate() const;
+
+   template<typename Stream>
+   friend Stream& operator << ( Stream& out, const chain_config_v2& c ) {
+      return c.log(out) << "\n";
+   }
+
+   friend inline bool operator == ( const chain_config_v2& lhs, const chain_config_v2& rhs ) {
+      //add v1 parameters comarison here
+      return ( std::tie(lhs.max_sync_call_depth, lhs.max_sync_call_data_size) ==
+               std::tie(lhs.max_sync_call_depth, rhs.max_sync_call_data_size) ) &&
+             lhs.base() == rhs.base();
+   }
+
+   friend inline bool operator != ( const chain_config_v2& lhs, const chain_config_v2& rhs ) {
+      return !(lhs == rhs);
+   }
+
+   inline chain_config_v2& operator= (const chain_config_v0& b) {
+      chain_config_v0::operator= (b);
+      return *this;
+   }
```

Comment:
Why do you need an assignment operator from v0 that leaves v1 and v2 data alone? If this is needed, seems like an explicit `copy()` method would be safer.

Reply:
This is required by https://github.com/AntelopeIO/spring/blob/2ffd8980a1566054bded1d55d44d3d388b03d3bf/libraries/chain/controller.cpp#L2568

Reply:
Actually there are 3 places using it.

Done by https://github.com/AntelopeIO/spring/pull/1257/commits/1890e98892e07c274bb2d434ca2a3c15f6c87ca5

Reply:
What makes it safer? At least the old way didn't have to repeat every item manually (which seems _less_ safe)

Reply:
I thought about that too. The new version explicitly sets v1 and v2. But that could be done without manually setting the long list of v0 fields. Should I revert to the original version, only adding the v1 and v2 portion?

Reply:
If we don't like the unusual `operator=` aspect of this, we could still have a `copy()` that called `chain_config_v0::operator= (other)` I think? That way you don't have the unusual `operator=` and don't need to repeat each v0 item (personally I'd suggest we def avoid that). This is all mostly a style thing though; the original unusual `operator=` doesn't bother me all that much (esp now that we understand why it's there), so need to hear more @heifner thoughts on it.

Reply:
I like this compromise.

Reply:
https://github.com/AntelopeIO/spring/pull/1257/commits/7a19b34dc68cfe0c28f4e73ea5754385f9420ab1

Reply:
I like the compromise as well.